eth/eventservices/roundinitializer.go

```diff
@@ -1,63 +1,92 @@
-type BlockNumReader interface {
-	LastSeenBlock() (*big.Int, error)
-}
-
-// BlockHashReader describes methods for reading the last initialized block hash
-type BlockHashReader interface {
+type timeWatcher interface {
+	LastSeenBlock() *big.Int
+	LastInitializedRound() *big.Int
 	LastInitializedBlockHash() [32]byte
+	CurrentRoundStartBlock() *big.Int
+	SubscribeRounds(sink chan<- types.Log) event.Subscription
+	SubscribeBlocks(sink chan<- *big.Int) event.Subscription
+	GetTranscoderPoolSize() *big.Int
 }
 
 // RoundInitializer is a service that automatically initializes the current round. Each round is split into epochs with a length of
 // epochBlocks. During each epoch a member of the upcoming active set is selected to initialize the round
 // This selection process is purely a client side implementation that attempts to minimize on-chain transaction collisions, but
 // collisions are still possible if initialization transactions are submitted by parties that are not using this selection process
 type RoundInitializer struct {
-	client          LivepeerEthClient
-	blkNumRdr       BlockNumReader
-	blkHashRdr      BlockHashReader
-	pollingInterval time.Duration
+	client eth.LivepeerEthClient
+	tw     timeWatcher
+	quit   chan struct{}
 
-	quit chan struct{}
+	nextRoundStartBlock *big.Int
 }
 
 // NewRoundInitializer creates a RoundInitializer instance
-func NewRoundInitializer(client LivepeerEthClient, blkNumRdr BlockNumReader, blkHashRdr BlockHashReader, pollingInterval time.Duration) *RoundInitializer {
+func NewRoundInitializer(client eth.LivepeerEthClient, tw timeWatcher) *RoundInitializer {
 	return &RoundInitializer{
-		client:          client,
-		blkNumRdr:       blkNumRdr,
-		blkHashRdr:      blkHashRdr,
-		pollingInterval: pollingInterval,
-		quit:            make(chan struct{}),
+		client: client,
+		tw:     tw,
+		quit:   make(chan struct{}),
 	}
 }
 
 // Start kicks off a loop that checks if the round should be initialized
-func (r *RoundInitializer) Start() {
-	ticker := time.NewTicker(r.pollingInterval)
+func (r *RoundInitializer) Start() error {
+	blockSink := make(chan *big.Int, 10)
+	blockSub := r.tw.SubscribeBlocks(blockSink)
+	defer blockSub.Unsubscribe()
+
+	roundSink := make(chan types.Log, 10)
+	roundSub := r.tw.SubscribeRounds(roundSink)
+	defer roundSub.Unsubscribe()
+
+	roundLength, err := r.client.RoundLength()
+	if err != nil {
+		return err
+	}
+
+	currentRoundStartBlock, err := r.client.CurrentRoundStartBlock()
+	if err != nil {
+		return err
+	}
+
+	r.nextRoundStartBlock = new(big.Int).Add(currentRoundStartBlock, roundLength)
 
 	for {
 		select {
 		case <-r.quit:
-			ticker.Stop()
-			return
-		case <-ticker.C:
-			if err := r.tryInitialize(); err != nil {
-				glog.Errorf("error trying to initialize round: %v", err)
+			glog.Infof("Stopping round initializer")
+			return nil
+		case err := <-blockSub.Err():
+			if err != nil {
+				glog.Error(err)
```

Comment:
Annotate this error (and in the next case) too as is done in the reward service?

Reply:
fixed but already caught in the rebase